### PR TITLE
Memory Optimizations for Convolution and Batch Normalization Layers

### PR DESCRIPTION
**Pull request by fiedorowicz1**
Allow the convolution and batch normalization layers to reuse their inputs for their error signals if the parent layer does not need its activations for backprop.

### CODE REVIEW

**Review by tbennun — APPROVED** [review]

**Review by benson31 — APPROVED** [review]
LGTM

**Review by bvanessen — APPROVED** [review]
